File: 백준/Silver/2156. 포도주 시식/포도주 시식.py
import sys
input= sys.stdin.readline

n = int(input())
data = []
dp = [0]*n

for i in range(n):
    data.append(int(input()))

dp[0] = data[0]
if n > 1:
    dp[1] = data[0]+data[1]

if n > 2:
    dp[2] = max(data[2]+data[1], data[2]+data[0], dp[1])

for i in range(3, n):
    dp[i] = max(dp[i-1], dp[i-3]+data[i-1]+data[i], dp[i-2]+data[i])

print(dp[n-1])

File: 백준/Silver/2156. 포도주 시식/포도주 시식.java
import java.io.*;
import java.util.*;

public class Main {

    //연속으로 놓여있는 3잔을 모두 마실 수 없다. -> 이거 어떻게 해결할건데?
    public static void main(String[] args) throws IOException {
        //입력부
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int n = Integer.parseInt(br.readLine());
        int[] data = new int[n + 1];
        int[] dp = new int[n + 1];
        for(int i=1; i<=n; i++) {
            data[i] = Integer.parseInt(br.readLine());
        }

        //로직 시작
        dp[1] = data[1];
        if(n >= 2) {
            dp[2] = dp[1] + data[2];
        }

        for(int i=3; i<=n; i++) {
            dp[i] = Math.max(dp[i - 1], Math.max(dp[i - 2] + data[i], dp[i - 3] + data[i - 1] + data[i]));
        }

        System.out.println(Arrays.stream(dp).max().getAsInt());
    }
}
